packages/vats/src/lib-chainPublish.js

```diff
@@ -0,0 +1,46 @@
+// @ts-check
+
+import { E } from '@endo/far';
+
+/**
+ * Publish results of an async iterable into a chain storage node as an array
+ * serialized using JSON.stringify or an optionally provided function (e.g.,
+ * leveraging a serialize function from makeMarshal).
+ *
+ * @param {AsyncIterator} iterable
+ * @param {ReturnType<typeof import('./lib-chainStorage.js').makeChainStorageRoot>} chainStorageNode
+ * @param {{ timerService: ERef<TimerService>, serialize?: (obj: any) => string }} powers
+ */
+export async function publishToChainNode(
+  iterable,
+  chainStorageNode,
+  { timerService, serialize = JSON.stringify },
+) {
+  let nextIndex = 0n;
+  let isNewBlock = true;
+  let oldResults = [];
+  let results = [];
+  for await (const result of iterable) {
+    if (isNewBlock) {
+      isNewBlock = false;
+      oldResults = results;
+      results = [];
+      E(timerService)
+        .delay(1n)
+        .then(() => {
+          isNewBlock = true;
+        });
+    }
+    // To avoid loss when detecting the new block *after* already consuming
+    // results produced within it, we associate each result with an index and
+    // include "old" results in the batch.
+    // Downstream consumers are expected to deduplicate by index.
```

Comment:
I didn't see a way around this, because there's no way to know when the timerService delay resolves (or at any other time, for that matter) which results have actually been published in a block vs. which are still pending.

Reply:
Could you elaborate on how?